csharp/src/Ice/Communicator-OutgoingConnectionFactory.cs

```diff
@@ -0,0 +1,270 @@
+                        }
+                    }
+
+                    // If we didn't find an active connection check if there is a pending connect task for the same
+                    // endpoint and label.
+                    if (!_pendingOutgoingConnections.TryGetValue((endpoint, label), out connectTask))
+                    {
+                        connectTask = PerformConnectAsync(endpoint, preferNonSecure, label);
+                        if (!connectTask.IsCompleted)
+                        {
+                            // If the task didn't complete synchronously we add it to the pending map
+                            // and it will be removed once PerformConnectAsync completes.
+                            _pendingOutgoingConnections[(endpoint, label)] = connectTask;
+                        }
+                    }
+                }
+
+                connection = await connectTask.WaitAsync(cancel).ConfigureAwait(false);
+                // After the connect task completed check if the connection can be trusted.
+                if (!connection.CanTrust(preferNonSecure))
+                {
+                    // The connection cannot be trusted clear the connection and try again.
+                    connection = null;
+                }
+            }
+            while (connection == null);
+            Debug.Assert(preferNonSecure != NonSecure.Never ||
+                         connection is ColocatedConnection ||
+                         connection.IsSecure);
+            return connection;
+
+            async Task<Connection> PerformConnectAsync(
+                Endpoint endpoint,
+                NonSecure preferNonSecure,
+                object? label)
+            {
+                try
+                {
+                    Debug.Assert(ConnectTimeout > TimeSpan.Zero);
+                    using var source = new CancellationTokenSource(ConnectTimeout);
+                    using var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(
+                        source.Token,
+                        endpoint.Communicator.CancellationToken);
+                    CancellationToken cancel = linkedSource.Token;
+                    Connection connection = await endpoint.ConnectAsync(preferNonSecure,
+                                                                        label,
+                                                                        cancel).ConfigureAwait(false);
+                    lock (_mutex)
+                    {
+                        if (_disposeTask != null)
+                        {
+                            // If the communicator has been already disposed throw here and avoid adding the connection
+                            // to the outgoing connections map, the connection will be disposed from the pending
+                            // connections map.
+                            throw new CommunicatorDisposedException();
```

Comment:
We need to close the connection as well.

Reply:
do we? I think that if we are here it means connect didn't complete synchronously, and close will be called on the pending set.

Reply:
If you mean that `Communicator.DisposeAsync` will close the connection because it's in the pending set, I don't see how if we throw here. Or did you perhaps intend to return the connection in this case instead of throwing?

Reply:
You are right, I actually need to return the connection for that to work. Not sure what will be simpler more correct here?